### PR TITLE
add extract-cache and save-cache commands

### PR DESCRIPTION
**Pull request by netanelgilad**
Initial implementation of the cache commands for npm-ci.
Assumes AWS credentials exist on the build agent (works the same for other CI builds that use s3).

Design decisions:
* Based on `.ci_config`. We can come up with some heuristic on how to get the paths that need to be cached. I'd rather start with explicitness and if we are building upon the `.ci_config` file, I guess that is a good place to put it. (under `cache.paths`).
* The default cache key is going to be `ORG_REPO_BRANCH`. That means the cache is shared per branch build inside a repo. In pull requests, that means the first build of the branch will not have a cache, but subsequent builds will. (Maybe take the first cache from master? Check if the cache for the branch exists and fallback to master (and have master hard-coded as a base branch)? need to think about it). On long-lived branches, most of the builds will have the cache.

### CODE REVIEW

File: src/cache.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  }).createReadStream()
+    .pipe(tar.extract({}))
+    .on('error', err => console.log('error while downloading and extracting cache', err))
+    .on('end', () => {
+      console.log(`Finished downloading and extracting cache.`);
+    });
+}
+
+export async function saveCache() {
+  const ciConfig = JSON.parse(readFileSync('.ci_config', 'utf8'));
```
Comment (yairhaimo) [inline src/cache.js]: What happens if `.ci_config` doesnt exist?

File: src/cache.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
+const {readFileSync, createReadStream} = require('fs');
+const tar = require('tar');
+const AWS = require('aws-sdk');
+const {sync: globbySync} = require('globby');
+const tempy = require('tempy');
+
+AWS.config.credentials = new AWS.SharedIniFileCredentials({profile: 'automation-aws'});
```
Comment (yairhaimo) [inline src/cache.js]: Could this fail in any way?

File: src/cache.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+      expandDirectories: false
+    });
+
+    console.log('Expanded the globs to the following paths to cache:');
+    pathsToCache.forEach(console.log.bind(console));
+
+    const tempFile = tempy.file({extension: 'tar.gz'});
+
+    console.log(`Creating tar.gz of cache paths at ${tempFile}...`);
+    await tar.create({
```
Comment (yairhaimo) [inline src/cache.js]: Maybe should wrap in `try`/`catch` just in case

Comment (netanelgilad) [reply src/cache.js]: I was planning on wrapping both commands with `|| exit 0` in the `npmBuild.sh` so I could avoid any kind of error handling in these commands, but now that I checked, I'm not sure I need to call the commands from there. Will update.

**Comment by yairhaimo** [comment]
I understand you plan to use these commands from the `npmBuild.sh` file?
Note that you will need to add it in multiple places. Also a long term goal, in my opinion, is to reduce the amount logic inside the shell script.
A better option in my opinion would be to put it as part of the `install` command in `npm-ci-scripts`. WDYT?

**Comment by netanelgilad** [comment]
At first I thought about just wrapping the `install` command. But looking at the implementation of other CI providers, and a bit common sense, I think we need to save the cache only if the build succeeded.

I think the ideal solution would have to put in the build step in the CI, as an after success or similar (not sure if it's available). But since we don't want to require a change to the CI template and have control over it, I think I'll just do it in the `build` command, which seems to be the one that runs the main process of install -> build -> test, and I'll change it to extract-cache -> install -> build -> test -> save-cache.

**Pull request merged**